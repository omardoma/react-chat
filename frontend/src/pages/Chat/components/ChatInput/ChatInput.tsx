import React, { useState, useRef } from 'react';

import { useIntl } from 'react-intl';
import { makeStyles } from '@material-ui/core/styles';
import { Paper, Input, IconButton } from '@material-ui/core';
import { Send as SendIcon } from '@material-ui/icons';

const useStyles = makeStyles({
  root: {
    display: 'flex'
  },
  input: {
    flex: 1,
    padding: '0.5rem'
  },
  sendButton: {
    flexShrink: 0
  }
});

interface ChatInputProps {
  sendMessage: (text: string) => void;
  sendOnCtrlEnter: boolean;
}

function ChatInput(props: ChatInputProps): JSX.Element {
  const { formatMessage: f } = useIntl();
  const classes = useStyles();
  const { sendMessage, sendOnCtrlEnter } = props;
  const [text, setText] = useState<string>('');
  const sendButtonEl = useRef<HTMLButtonElement>(null);
  const handleInputChange = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
    setText(event.target.value);
  };
  const handleFormSubmit = (event: React.FormEvent<HTMLElement>) => {
    event.preventDefault();
    const trimmedText = text.trim();
    if (trimmedText) {
      sendMessage(trimmedText);
      setText('');
    }
  };
  const handleInputKeyPress = (event: React.KeyboardEvent<HTMLElement>) => {
    if (
      event.key === 'Enter' &&
      ((sendOnCtrlEnter && event.ctrlKey) ||
        !(sendOnCtrlEnter || event.ctrlKey))
    ) {
      event.preventDefault();
      sendButtonEl.current.click();
    }
  };
  return (
    <Paper
      component="form"
      variant="outlined"
      square
      className={classes.root}
      onSubmit={handleFormSubmit}
    >
      <Input
        className={classes.input}
        multiline
        rowsMax={6}
        autoComplete="off"
        autoFocus
        disableUnderline
        placeholder={f({ id: 'chat.chatInput.placeholder' })}
        inputProps={{ 'aria-label': f({ id: 'chat.chatInput.label' }) }}
        name="message"
        value={text}
        onChange={handleInputChange}
        onKeyPress={handleInputKeyPress}
      />
      <IconButton
        ref={sendButtonEl}
        type="submit"
        className={classes.sendButton}
        aria-label={f({ id: 'chat.chatInput.sendButton' })}
      >
        <SendIcon />
      </IconButton>
    </Paper>
  );
}

export default ChatInput;
