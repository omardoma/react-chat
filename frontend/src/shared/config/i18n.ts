export default {
  en: {
    'navbar.tabs.chat': 'Chat',
    'navbar.tabs.settings': 'Settings',
    'settings.form.username.label': 'Username',
    'settings.form.theme.label': 'Interface color',
    'settings.form.theme.options.light': 'Light',
    'settings.form.theme.options.dark': 'Dark',
    'settings.form.clockFormat.label': 'Clock display',
    'settings.form.clockFormat.options.hh': '12 Hours',
    'settings.form.clockFormat.options.HH': '24 Hours',
    'settings.form.sendOnCtrlEnter.label': 'Send messages on CTRL+ENTER',
    'settings.form.sendOnCtrlEnter.options.on': 'On',
    'settings.form.sendOnCtrlEnter.options.off': 'Off',
    'settings.form.language.label': 'Language',
    'settings.form.resetButton': 'Reset to defaults',
    'chat.chatInput.label': 'Message Input',
    'chat.chatInput.placeholder': 'Enter Message',
    'chat.chatInput.sendButton': 'Send',
    'chat.emptyMessage': 'No messages, start the chat!',
    newMessage: 'You have a new message!'
  },
  de: {
    'navbar.tabs.chat': 'Plaudern',
    'navbar.tabs.settings': 'die Einstellungen',
    'settings.form.username.label': 'Nutzername',
    'settings.form.theme.label': 'Schnittstellenfarbe',
    'settings.form.theme.options.light': 'Licht',
    'settings.form.theme.options.dark': 'dunkel',
    'settings.form.clockFormat.label': 'Uhrenanzeige',
    'settings.form.clockFormat.options.hh': '12 Stunden',
    'settings.form.clockFormat.options.HH': '24 Stunden',
    'settings.form.sendOnCtrlEnter.label':
      'Senden Sie Nachrichten mit STRG + EINGABETASTE',
    'settings.form.sendOnCtrlEnter.options.on': 'Auf',
    'settings.form.sendOnCtrlEnter.options.off': 'aus',
    'settings.form.language.label': 'Sprache',
    'settings.form.resetButton': 'Auf Standardeinstellung zurücksetzen',
    'chat.chatInput.label': 'Nachrichteneingabe',
    'chat.chatInput.placeholder': 'Nachricht eingeben',
    'chat.chatInput.sendButton': 'Senden',
    'chat.emptyMessage': 'Keine Nachrichten, starte das Gespräch!',
    newMessage: 'Du hast eine neue Nachricht!'
  }
};
