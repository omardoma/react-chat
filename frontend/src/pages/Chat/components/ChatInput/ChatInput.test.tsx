import React from 'react';

import { render, screen, userEvent, fireEvent } from 'test-utils';

import ChatInput from './ChatInput';

import i18n from '../../../../shared/config/i18n';

describe('ChatInput', () => {
  it('renders ChatInput component', () => {
    // Arrange
    const sendMessage = jest.fn();
    const sendOnCtrlEnter = false;
    const { asFragment } = render(
      <ChatInput sendMessage={sendMessage} sendOnCtrlEnter={sendOnCtrlEnter} />
    );
    // Assert
    expect(asFragment()).toMatchSnapshot();
  });

  it('should type the user input', () => {
    // Arrange
    const sendMessage = jest.fn();
    const sendOnCtrlEnter = false;
    render(
      <ChatInput sendMessage={sendMessage} sendOnCtrlEnter={sendOnCtrlEnter} />
    );
    const messageInputEl = screen.getByRole('textbox', { name: /message/i });
    const messageText = 'Hey there';
    // Act
    userEvent.type(messageInputEl, messageText);
    // Assert
    expect(messageInputEl).toHaveValue(messageText);
  });

  it('should send a message when the send button is clicked', () => {
    // Arrange
    const sendMessage = jest.fn();
    const sendOnCtrlEnter = false;
    render(
      <ChatInput sendMessage={sendMessage} sendOnCtrlEnter={sendOnCtrlEnter} />
    );
    const messageInputEl = screen.getByRole('textbox', { name: /message/i });
    const sendButtonEl = screen.getByLabelText(
      i18n.en['chat.chatInput.sendButton']
    );
    const messageText = 'Hey there';
    // Act
    userEvent.type(messageInputEl, messageText);
    userEvent.click(sendButtonEl);
    // Assert
    expect(sendMessage).toHaveBeenCalledTimes(1);
    expect(sendMessage).toHaveBeenCalledWith(messageText);
  });

  it('should not send a message when the Enter key is pressed and sendOnCtrlEnter is on', () => {
    // Arrange
    const sendMessage = jest.fn();
    const sendOnCtrlEnter = true;
    render(
      <ChatInput sendMessage={sendMessage} sendOnCtrlEnter={sendOnCtrlEnter} />
    );
    const messageInputEl = screen.getByRole('textbox', { name: /message/i });
    const messageText = 'Hey there';
    // Act
    userEvent.type(messageInputEl, messageText);
    fireEvent.keyPress(messageInputEl, {
      key: 'Enter',
      code: 13,
      charCode: 13,
      keyCode: 13
    });
    // Assert
    expect(sendMessage).not.toHaveBeenCalled();
  });

  it('should send a message when the Enter key is pressed and sendOnCtrlEnter is off', () => {
    // Arrange
    const sendMessage = jest.fn();
    const sendOnCtrlEnter = false;
    render(
      <ChatInput sendMessage={sendMessage} sendOnCtrlEnter={sendOnCtrlEnter} />
    );
    const messageInputEl = screen.getByRole('textbox', { name: /message/i });
    const messageText = 'Hey there';
    // Act
    userEvent.type(messageInputEl, messageText);
    fireEvent.keyPress(messageInputEl, {
      key: 'Enter',
      code: 13,
      charCode: 13,
      keyCode: 13
    });
    // Assert
    expect(sendMessage).toHaveBeenCalledTimes(1);
    expect(sendMessage).toHaveBeenCalledWith(messageText);
  });

  it('should not send a message when CTRL+ENTER key combination is pressed and sendOnCtrlEnter is off', () => {
    // Arrange
    const sendMessage = jest.fn();
    const sendOnCtrlEnter = false;
    render(
      <ChatInput sendMessage={sendMessage} sendOnCtrlEnter={sendOnCtrlEnter} />
    );
    const messageInputEl = screen.getByRole('textbox', { name: /message/i });
    const messageText = 'Hey there';
    // Act
    userEvent.type(messageInputEl, messageText);
    fireEvent.keyPress(messageInputEl, {
      key: 'Enter',
      code: 13,
      charCode: 13,
      keyCode: 13,
      ctrlKey: true
    });
    // Assert
    expect(sendMessage).not.toHaveBeenCalled();
  });

  it('should send a message when CTRL+ENTER key combination is pressed and sendOnCtrlEnter is on', () => {
    // Arrange
    const sendMessage = jest.fn();
    const sendOnCtrlEnter = true;
    render(
      <ChatInput sendMessage={sendMessage} sendOnCtrlEnter={sendOnCtrlEnter} />
    );
    const messageInputEl = screen.getByRole('textbox', { name: /message/i });
    const messageText = 'Hey there';
    // Act
    userEvent.type(messageInputEl, messageText);
    fireEvent.keyPress(messageInputEl, {
      key: 'Enter',
      code: 13,
      charCode: 13,
      keyCode: 13,
      ctrlKey: true
    });
    // Assert
    expect(sendMessage).toHaveBeenCalledTimes(1);
    expect(sendMessage).toHaveBeenCalledWith(messageText);
  });
});
