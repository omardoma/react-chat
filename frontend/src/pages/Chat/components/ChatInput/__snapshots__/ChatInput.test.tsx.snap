// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatInput renders ChatInput component 1`] = `
<DocumentFragment>
  <form
    class="MuiPaper-root makeStyles-root-1 MuiPaper-outlined"
  >
    <div
      class="MuiInputBase-root MuiInput-root makeStyles-input-2 Mui-focused Mui-focused MuiInputBase-multiline MuiInput-multiline"
    >
      <textarea
        aria-label="Message Input"
        autocomplete="off"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline"
        name="message"
        placeholder="Enter Message"
        rows="1"
        style="height: 0px; overflow: hidden;"
      />
      <textarea
        aria-hidden="true"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline"
        readonly=""
        style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;"
        tabindex="-1"
      />
    </div>
    <button
      aria-label="Send"
      class="MuiButtonBase-root MuiIconButton-root makeStyles-sendButton-3"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M2.01 21L23 12 2.01 3 2 10l15 2-15 2z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;
